### PR TITLE
Feature/fix storagemanager layout size

### PR DESCRIPTION
**Pull request by khancyr**
fix https://github.com/ArduPilot/ardupilot/issues/10389
use compile time option to define storage layout. It solve the issue and reduce the build size.
I also remove the memset as direct initialisation to zero is faster in this case and gain some size.
I move the define to the cpp to reduce scope

### CODE REVIEW

**Comment by peterbarker** [comment]
Isn't this going to stuff people's existing storage?  You're effectively modifying the layout.

@WickedShell was musing recently we should revisit the way we do storage generally.

**Comment by khancyr** [comment]
it there a way to check how the layout is ? normally the layout is unchanged, it just correct the storage size reporting that is only calculated on instantiation and then before swtiching to copter layout.

**Comment by peterbarker** [comment]
OK, now I've actually gone back and had a thorough read I think I like this.

I haven't tested it yet.

File: libraries/StorageManager/StorageManager.cpp [hunk 1]
```diff
@@ -35,18 +52,24 @@ extern const AP_HAL::HAL& hal;
   On PX4v1 this gives 309 waypoints, 30 rally points and 52 fence points
   On Pixhawk this gives 724 waypoints, 50 rally points and 84 fence points
  */
-const StorageManager::StorageArea StorageManager::layout_default[STORAGE_NUM_AREAS] = {
```
Comment (peterbarker) [inline libraries/StorageManager/StorageManager.cpp]: We've lost the sizing here, which I like, but should we have a `static_assert()` here somewhere to ensure the layout is the size we expect?

Comment (khancyr) [reply libraries/StorageManager/StorageManager.cpp]: Yes ,we should have a static_assert.I always forget about those ...

Comment (khancyr) [reply libraries/StorageManager/StorageManager.cpp]: done !

**Comment by khancyr** [comment]
rebase and add static_assert

**Comment by lvale** [comment]
How does this relate to https://github.com/ArduPilot/ardupilot/pull/4172 ?

**Comment by tridge** [comment]
pending testing on a board with OSD param storage

**Comment by khancyr** [comment]
@tridge hello, I don't have Osb board. But we should be able to test on SITL, right ? what do you want to test exactly ?

**Comment by khancyr** [comment]
WITH FIX : 
On SITL with OSD , copter : 
````
i: 0, offset: 0, length: 1536
i: 1, offset: 1536, length: 2422
i: 2, offset: 3958, length: 90
i: 3, offset: 4048, length: 48
i: 4, offset: 4096, length: 1280
i: 5, offset: 5376, length: 300
i: 6, offset: 5676, length: 256
i: 7, offset: 5932, length: 2132
i: 8, offset: 8064, length: 64
i: 9, offset: 8128, length: 56
i: 10, offset: 8192, length: 1280
i: 11, offset: 9472, length: 300
i: 12, offset: 9772, length: 256
i: 13, offset: 10028, length: 6228
type: 3, total size 10782, total area 14
i: 0, offset: 0, length: 1536
i: 1, offset: 1536, length: 2422
i: 2, offset: 3958, length: 90
i: 3, offset: 4048, length: 48
i: 4, offset: 4096, length: 1280
i: 5, offset: 5376, length: 300
i: 6, offset: 5676, length: 256
i: 7, offset: 5932, length: 2132
i: 8, offset: 8064, length: 64
i: 9, offset: 8128, length: 56
i: 10, offset: 8192, length: 1280
i: 11, offset: 9472, length: 300
i: 12, offset: 9772, length: 256
i: 13, offset: 10028, length: 6228
type: 0, total size 4096, total area 14
i: 0, offset: 0, length: 1536
i: 1, offset: 1536, length: 2422
i: 2, offset: 3958, length: 90
i: 3, offset: 4048, length: 48
i: 4, offset: 4096, length: 1280
i: 5, offset: 5376, length: 300
i: 6, offset: 5676, length: 256
i: 7, offset: 5932, length: 2132
i: 8, offset: 8064, length: 64
i: 9, offset: 8128, length: 56
i: 10, offset: 8192, length: 1280
i: 11, offset: 9472, length: 300
i: 12, offset: 9772, length: 256
i: 13, offset: 10028, length: 6228
type: 2, total size 690, total area 14
i: 0, offset: 0, length: 1536
i: 1, offset: 1536, length: 2422
i: 2, offset: 3958, length: 90
i: 3, offset: 4048, length: 48
i: 4, offset: 4096, length: 1280
i: 5, offset: 5376, length: 300
i: 6, offset: 5676, length: 256
i: 7, offset: 5932, length: 2132
i: 8, offset: 8064, length: 64
i: 9, offset: 8128, length: 56
i: 10, offset: 8192, length: 1280
i: 11, offset: 9472, length: 300
i: 12, offset: 9772, length: 256
i: 13, offset: 10028, length: 6228
type: 4, total size 64, total area 14
i: 0, offset: 0, length: 1536
i: 1, offset: 1536, length: 2422
i: 2, offset: 3958, length: 90
i: 3, offset: 4048, length: 48
i: 4, offset: 4096, length: 1280
i: 5, offset: 5376, length: 300
i: 6, offset: 5676, length: 256
i: 7, offset: 5932, length: 2132
i: 8, offset: 8064, length: 64
i: 9, offset: 8128, length: 56
i: 10, offset: 8192, length: 1280
i: 11, offset: 9472, length: 300
i: 12, offset: 9772, length: 256
i: 13, offset: 10028, length: 6228
type: 1, total size 560, total area 14
````

On rover
````
i: 5, offset: 5376, length: 300
i: 6, offset: 5676, length: 256
i: 7, offset: 5932, length: 2132
i: 8, offset: 8064, length: 64
i: 9, offset: 8128, length: 56
i: 10, offset: 8192, length: 1280
i: 11, offset: 9472, length: 300
i: 12, offset: 9772, length: 256
i: 13, offset: 10028, length: 6228
type: 3, total size 10866, total area 14
i: 0, offset: 0, length: 1280
i: 1, offset: 1280, length: 2506
i: 2, offset: 3786, length: 150
i: 3, offset: 3936, length: 160
i: 4, offset: 4096, length: 1280
i: 5, offset: 5376, length: 300
i: 6, offset: 5676, length: 256
i: 7, offset: 5932, length: 2132
i: 8, offset: 8064, length: 64
i: 9, offset: 8128, length: 56
i: 10, offset: 8192, length: 1280
i: 11, offset: 9472, length: 300
i: 12, offset: 9772, length: 256
i: 13, offset: 10028, length: 6228
type: 0, total size 3840, total area 14
i: 0, offset: 0, length: 1280
i: 1, offset: 1280, length: 2506
i: 2, offset: 3786, length: 150
i: 3, offset: 3936, length: 160
i: 4, offset: 4096, length: 1280
i: 5, offset: 5376, length: 300
i: 6, offset: 5676, length: 256
i: 7, offset: 5932, length: 2132
i: 8, offset: 8064, length: 64
i: 9, offset: 8128, length: 56
i: 10, offset: 8192, length: 1280
i: 11, offset: 9472, length: 300
i: 12, offset: 9772, length: 256
i: 13, offset: 10028, length: 6228
type: 2, total size 750, total area 14
i: 0, offset: 0, length: 1280
i: 1, offset: 1280, length: 2506
i: 2, offset: 3786, length: 150
i: 3, offset: 3936, length: 160
i: 4, offset: 4096, length: 1280
i: 5, offset: 5376, length: 300
i: 6, offset: 5676, length: 256
i: 7, offset: 5932, length: 2132
i: 8, offset: 8064, length: 64
i: 9, offset: 8128, length: 56
i: 10, offset: 8192, length: 1280
i: 11, offset: 9472, length: 300
i: 12, offset: 9772, length: 256
i: 13, offset: 10028, length: 6228
type: 4, total size 64, total area 14
i: 0, offset: 0, length: 1280
i: 1, offset: 1280, length: 2506
i: 2, offset: 3786, length: 150
i: 3, offset: 3936, length: 160
i: 4, offset: 4096, length: 1280
i: 5, offset: 5376, length: 300
i: 6, offset: 5676, length: 256
i: 7, offset: 5932, length: 2132
i: 8, offset: 8064, length: 64
i: 9, offset: 8128, length: 56
i: 10, offset: 8192, length: 1280
i: 11, offset: 9472, length: 300
i: 12, offset: 9772, length: 256
i: 13, offset: 10028, length: 6228
type: 1, total size 672, total area 14
````

On master
copter : 
````
i: 5, offset: 5376, length: 300
i: 6, offset: 5676, length: 256
i: 7, offset: 5932, length: 2132
i: 8, offset: 8064, length: 64
i: 9, offset: 8128, length: 56
i: 10, offset: 8192, length: 1280
i: 11, offset: 9472, length: 300
i: 12, offset: 9772, length: 256
i: 13, offset: 10028, length: 6228
type: 3, total size 10866, total area 14
i: 0, offset: 0, length: 1280
i: 1, offset: 1280, length: 2506
i: 2, offset: 3786, length: 150
i: 3, offset: 3936, length: 160
i: 4, offset: 4096, length: 1280
i: 5, offset: 5376, length: 300
i: 6, offset: 5676, length: 256
i: 7, offset: 5932, length: 2132
i: 8, offset: 8064, length: 64
i: 9, offset: 8128, length: 56
i: 10, offset: 8192, length: 1280
i: 11, offset: 9472, length: 300
i: 12, offset: 9772, length: 256
i: 13, offset: 10028, length: 6228
type: 0, total size 3840, total area 14
i: 0, offset: 0, length: 1280
i: 1, offset: 1280, length: 2506
i: 2, offset: 3786, length: 150
i: 3, offset: 3936, length: 160
i: 4, offset: 4096, length: 1280
i: 5, offset: 5376, length: 300
i: 6, offset: 5676, length: 256
i: 7, offset: 5932, length: 2132
i: 8, offset: 8064, length: 64
i: 9, offset: 8128, length: 56
i: 10, offset: 8192, length: 1280
i: 11, offset: 9472, length: 300
i: 12, offset: 9772, length: 256
i: 13, offset: 10028, length: 6228
type: 2, total size 750, total area 14
i: 0, offset: 0, length: 1280
i: 1, offset: 1280, length: 2506
i: 2, offset: 3786, length: 150
i: 3, offset: 3936, length: 160
i: 4, offset: 4096, length: 1280
i: 5, offset: 5376, length: 300
i: 6, offset: 5676, length: 256
i: 7, offset: 5932, length: 2132
i: 8, offset: 8064, length: 64
i: 9, offset: 8128, length: 56
i: 10, offset: 8192, length: 1280
i: 11, offset: 9472, length: 300
i: 12, offset: 9772, length: 256
i: 13, offset: 10028, length: 6228
type: 4, total size 64, total area 14
i: 0, offset: 0, length: 1280
i: 1, offset: 1280, length: 2506
i: 2, offset: 3786, length: 150
i: 3, offset: 3936, length: 160
i: 4, offset: 4096, length: 1280
i: 5, offset: 5376, length: 300
i: 6, offset: 5676, length: 256
i: 7, offset: 5932, length: 2132
i: 8, offset: 8064, length: 64
i: 9, offset: 8128, length: 56
i: 10, offset: 8192, length: 1280
i: 11, offset: 9472, length: 300
i: 12, offset: 9772, length: 256
i: 13, offset: 10028, length: 6228
type: 1, total size 672, total area 14
````

Rover : 
````
i: 5, offset: 5376, length: 300
i: 6, offset: 5676, length: 256
i: 7, offset: 5932, length: 2132
i: 8, offset: 8064, length: 64
i: 9, offset: 8128, length: 56
i: 10, offset: 8192, length: 1280
i: 11, offset: 9472, length: 300
i: 12, offset: 9772, length: 256
i: 13, offset: 10028, length: 6228
type: 3, total size 10866, total area 14
i: 0, offset: 0, length: 1280
i: 1, offset: 1280, length: 2506
i: 2, offset: 3786, length: 150
i: 3, offset: 3936, length: 160
i: 4, offset: 4096, length: 1280
i: 5, offset: 5376, length: 300
i: 6, offset: 5676, length: 256
i: 7, offset: 5932, length: 2132
i: 8, offset: 8064, length: 64
i: 9, offset: 8128, length: 56
i: 10, offset: 8192, length: 1280
i: 11, offset: 9472, length: 300
i: 12, offset: 9772, length: 256
i: 13, offset: 10028, length: 6228
type: 0, total size 3840, total area 14
i: 0, offset: 0, length: 1280
i: 1, offset: 1280, length: 2506
i: 2, offset: 3786, length: 150
i: 3, offset: 3936, length: 160
i: 4, offset: 4096, length: 1280
i: 5, offset: 5376, length: 300
i: 6, offset: 5676, length: 256
i: 7, offset: 5932, length: 2132
i: 8, offset: 8064, length: 64
i: 9, offset: 8128, length: 56
i: 10, offset: 8192, length: 1280
i: 11, offset: 9472, length: 300
i: 12, offset: 9772, length: 256
i: 13, offset: 10028, length: 6228
type: 2, total size 750, total area 14
i: 0, offset: 0, length: 1280
i: 1, offset: 1280, length: 2506
i: 2, offset: 3786, length: 150
i: 3, offset: 3936, length: 160
i: 4, offset: 4096, length: 1280
i: 5, offset: 5376, length: 300
i: 6, offset: 5676, length: 256
i: 7, offset: 5932, length: 2132
i: 8, offset: 8064, length: 64
i: 9, offset: 8128, length: 56
i: 10, offset: 8192, length: 1280
i: 11, offset: 9472, length: 300
i: 12, offset: 9772, length: 256
i: 13, offset: 10028, length: 6228
type: 4, total size 64, total area 14
i: 0, offset: 0, length: 1280
i: 1, offset: 1280, length: 2506
i: 2, offset: 3786, length: 150
i: 3, offset: 3936, length: 160
i: 4, offset: 4096, length: 1280
i: 5, offset: 5376, length: 300
i: 6, offset: 5676, length: 256
i: 7, offset: 5932, length: 2132
i: 8, offset: 8064, length: 64
i: 9, offset: 8128, length: 56
i: 10, offset: 8192, length: 1280
i: 11, offset: 9472, length: 300
i: 12, offset: 9772, length: 256
i: 13, offset: 10028, length: 6228
type: 1, total size 672, total area 14
````

**Comment by khancyr** [comment]
@tridge ping on this one !

**Comment by khancyr** [comment]
Closed by https://github.com/ArduPilot/ardupilot/issues/13299

**Pull request closed without merging**